libraries/wasm/src/doc.rs

```diff
@@ -0,0 +1,260 @@
+    pub fn verify(&self) -> bool {
+        self.0.verify().is_ok()
+    }
+
+    #[wasm_bindgen]
+    pub fn diff(&self, other: &Doc, key: &Key) -> Result<JsValue, JsValue> {
+        let doc = DIDDocument::from_str(&other.0.to_string()).map_err(js_err)?;
+        let diff: DIDDiff = self.0.diff(doc, key.0.secret()).map_err(js_err)?;
+
+        JsValue::from_serde(&diff).map_err(js_err)
+    }
+
+    #[wasm_bindgen]
+    pub fn verify_diff(&self, diff: String) -> bool {
+        match DIDDiff::from_json(&diff) {
+            Ok(diff) => self.0.verify_diff(&diff).is_ok(),
+            Err(_) => false,
+        }
+    }
+
+    #[wasm_bindgen]
+    pub fn update_service(&mut self, did: WasmDID, url: String, service_type: String) -> Result<Doc, JsValue> {
+        let endpoint = ServiceEndpoint::Url(url.parse().map_err(js_err)?);
+        let service = ServiceBuilder::default()
+            .id(DID::parse(did.0).map_err(js_err)?)
+            .service_type(service_type)
+            .endpoint(endpoint)
+            .build()
+            .map_err(js_err)?;
+        self.0.update_service(service);
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn clear_services(&mut self) -> Result<Doc, JsValue> {
+        self.0.clear_services();
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_public_key(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_public_key(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_auth(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_auth(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_assert(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_assert(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_verification(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_verification(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_delegation(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_delegation(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_invocation(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_invocation(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_agreement(&mut self, public_key: &PubKey) -> Result<Doc, JsValue> {
+        self.0.update_agreement(public_key.0.clone());
+        Ok(Doc(self.0.clone()))
+    }
+
+    #[wasm_bindgen]
+    pub fn update_time(&mut self) {
+        self.0.update_time();
+    }
+
+    #[wasm_bindgen]
+    pub fn resolve_key(&mut self, key_relation: &str) -> Result<PubKey, JsValue> {
+        let relation = match key_relation {
```

Comment:
Should be able to use `serde` for this if you add `#[derive(Deserialize)]` to `KeyRelation`

Reply:
How should `serde` be used there then?

Reply:
I should have suggested `FromStr` - it would look like this:

```rust
use core::str::FromStr;

impl FromStr for KeyRelation {
  type Err = crate::Error;

  fn from_str(string: &str) -> Result<Self, Self::Err> {
    match string {
      "VerificationMethod" => Ok(Self::VerificationMethod),
      "Authentication" => Ok(Self::Authentication),
      "AssertionMethod" => Ok(Self::AssertionMethod),
      "KeyAgreement" => Ok(Self::KeyAgreement),
      "CapabilityInvocation" => Ok(Self::CapabilityInvocation),
      "CapabilityDelegation" => Ok(Self::CapabilityDelegation),
      _ => Err(...)
    }
  }
}
```

and you would use it like:

```rust
let relation: KeyRelation = key_relation.parse()?;
```